src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -48,32 +51,34 @@ private class AggregatingTagger : ForegroundThreadAffinitizedObject, IAccurateTa
                 // from the owner of that diagnostic update.
                 _owner._diagnosticService.DiagnosticsUpdated += OnDiagnosticsUpdated;
 
-                // Kick off a background task to collect the initial set of diagnostics.
-                var cancellationToken = _initialDiagnosticsCancellationSource.Token;
-                var asyncToken = _owner._listener.BeginAsyncOperation(GetType() + ".GetInitialDiagnostics");
-                var task = Task.Run(() => GetInitialDiagnostics(cancellationToken), cancellationToken);
-                task.CompletesAsyncOperation(asyncToken);
+                // get teh initial set of diagnostics from the service.  Do this now (without
+                // kicking off a task) to prevent race conditions.  We don't want to get an 
+                // event from the diagnostic service that we then override with the set of
+                // diagnostics we get when we call GetInitialDiagnostics
+                GetInitialDiagnostics();
             }
 
-            private void GetInitialDiagnostics(CancellationToken cancellationToken)
+            private void GetInitialDiagnostics()
             {
-                cancellationToken.ThrowIfCancellationRequested();
+                this.AssertIsForeground();
 
                 // Also, collect the initial set of diagnostics to show.
                 var document = _subjectBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
+                _currentDocumentId = document?.Id;
+
                 if (document != null)
                 {
                     var project = document.Project;
                     var workspace = project.Solution.Workspace;
-                    foreach (var updateArgs in _owner._diagnosticService.GetDiagnosticsUpdatedEventArgs(workspace, project.Id, document.Id, cancellationToken))
+                    foreach (var updateArgs in _owner._diagnosticService.GetDiagnosticsUpdatedEventArgs(workspace, project.Id, document.Id, CancellationToken.None))
```

Comment:
Does this block until diagnostics are computed, or just return the set the service currently has computed?